subsys/net/lib/latmon/latmon.c

```diff
@@ -0,0 +1,563 @@
+	}
+
+	if (msg->measure_func(delta) < 0) {
+		if (data->overruns++ > conf->max_samples / 2) {
+			return -1;
+		}
+		/* Just an overrun */
+		return 1;
+	}
+	return 0;
+}
+
+static void monitor_thread_func(void *p1, void *p2, void *p3)
+{
+	struct latmon_message *msg = p1;
+	struct latmus_conf *conf = p2;
+	struct latmon_data *data = p3;
+	uint32_t delta = 0;
+	int ret = 0;
+
+	LOG_INF("Monitor thread priority: %d", MONITOR_THREAD_PRIORITY);
+
+	/* Prepare transfer thread */
+	start_xfer_thread(&msg->latmus);
+
+	LOG_INF("\tmonitoring started:");
+	LOG_INF("\t - samples per period: %u", conf->max_samples);
+	LOG_INF("\t - period: %u usecs", conf->period);
+	LOG_INF("\t - histogram cells: %u", conf->cells);
+
+	/* Sampling loop */
+	memset(data, 0, sizeof(*data));
+	data->min_lat = UINT32_MAX;
+	data->warmed = false;
+	delta = 0;
+	while (!abort_monitor) {
+		for (;;) {
+			ret = measure(&delta, msg, data, conf);
+			if (ret) {
+				if (ret < 0) {
+					LOG_ERR("\tExcessive overruns, abort!");
+					goto out;
+				}
+				continue;
+			}
+
+			if (!prepare_sample_data(delta, conf, data)) {
+				/* Sample data ready */
+				break;
+			}
+		}
+
+		if (enqueue_sample_data(data) < 0) {
+			break;
+		}
+	};
+out:
+	abort_xfer_thread();
+	k_sem_give(&monitor_done);
+	monitor_tid = NULL;
+
+	LOG_INF("\tmonitoring stopped");
+}
+
+static int broadcast_ip_address(struct in_addr *ip_addr)
+{
+	char ip_str[NET_IPV4_ADDR_LEN];
+	struct sockaddr_in broadcast;
+	int sock = -1;
+	int ret = -1;
+
+	sock = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("failed to create broadcast socket : %d", errno);
+		return -1;
+	}
+
+	broadcast.sin_addr.s_addr = htonl(INADDR_BROADCAST);
+	broadcast.sin_port = htons(LATMON_NET_PORT);
+	broadcast.sin_family = AF_INET;
+
+	net_addr_ntop(AF_INET, ip_addr, ip_str, sizeof(ip_str));
+
+	ret = zsock_sendto(sock, ip_str, strlen(ip_str), 0,
+			(struct sockaddr *)&broadcast, sizeof(broadcast));
+
+	zsock_close(sock);
+
+	return ret;
+}
+
+/* Get a socket to listen to Latmus requests */
+int net_latmon_get_socket(struct sockaddr_in *connection_addr)
+{
+	struct sockaddr_in addr = {
+		.sin_family = AF_INET,
+		.sin_addr.s_addr = htonl(INADDR_ANY),
+		.sin_port = htons(LATMON_NET_PORT)
+	};
+	int s, on = 1;
+
+	if (connection_addr) {
+		memcpy(&addr, connection_addr, sizeof(addr));
+	}
+
+	s = zsock_socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (s < 0) {
+		LOG_ERR("failed to create latmon socket : %d", errno);
+		return -1;
+	}
+
+	zsock_setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+	if (zsock_bind(s, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+		LOG_ERR("failed to bind latmon socket : %d", errno);
+		zsock_close(s);
+		return -1;
+	}
+
+	if (zsock_listen(s, 1) < 0) {
+		LOG_ERR("failed to listen on latmon socket : %d", errno);
+		zsock_close(s);
+		return -1;
+	}
+
+	return s;
+}
+
+/* Waits for connection from Latmus */
+int net_latmon_connect(int socket, struct in_addr *ip)
```

Comment:
And here if we change this to `net_latmon_connect(int socket, struct sockaddr *ip)`, then there is no need to change the API to support IPv6